packages/jest-circus/src/deadlineTimeout.ts

```diff
@@ -0,0 +1,35 @@
+import {getState} from './state';
+
+export async function withinDeadline<T>(promise: Promise<T>): Promise<T> {
+  const deadline = getState()?.currentlyRunningTest?.deadline;
+  if (undefined === deadline) {
+    throw new Error('bug! no deadline available');
+  }
+  return timeout(promise, deadline - Date.now());
+}
+
+function isUs(line: string): boolean {
+  return line.includes('deadlineTimeout') && line.includes('jest-circus');
+}
+
+async function timeout<T>(promise: Promise<T>, ms: number): Promise<T> {
+  let timeoutId;
+  try {
+    return await Promise.race([
+      promise,
+      (async () => {
+        await new Promise(resolve => {
+          timeoutId = setTimeout(resolve, ms);
+        });
+        const here = new Error(`deadline exceeded (waited here for ${ms}ms)`);
+        here.stack = here.stack
```

Comment:
should probably guard this behind some check that it actually timed out as accessing the stack is not free

Reply:
I believe this won't run unless the `setTimeout` "sleep" above resolves (it's awaited), which doesn't happen in the non-timeout case?

The promise will never resolve most of the time, though. My node isn't quite good enough to know if that's a real problem; whether we leak the closure and some context in it somehow.

Reply:
you're correct! I do think we'll be leaking as you mention, though.

Could we do something like

```ts
async function timeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  let timeoutId: ReturnType<typeof setTimeout> | undefined;
  let resolvePromise: () => void;

  try {
    return await Promise.race([
      promise,
      new Promise<void>((resolve, reject) => {
        resolvePromise = resolve;

        timeoutId = setTimeout(() => {
          timeoutId = undefined;

          const here = new Error(`deadline exceeded (waited here for ${ms}ms)`);
          here.stack = here.stack
            ?.split('\n')
            .filter(line => !isUs(line))
            .join('\n');

          reject(here);
        }, ms);
      })(),
    ]);
  } finally {
    resolvePromise.?();
    if (timeoutId) {
      clearTimeout(timeoutId);
    }
  }
}
```
?

Reply:
Apparently not. If you reject/throw from within the Promise there, you lose the stack trace, which messes up the whole purpose of this. `p-timeout` has the same bug.

I rewrote the old implementation to await something more like your code, but with some booleans for whether it resolved or was cancelled. (pushed)